File: mockspresso-easymock-powermock/src/main/java/com/episode6/hackit/mockspresso/easymock/powermock/EasyPowerMockMockerConfig.java
package com.episode6.hackit.mockspresso.easymock.powermock;

import com.episode6.hackit.mockspresso.api.MockerConfig;
import com.episode6.hackit.mockspresso.reflect.ReflectUtil;
import com.episode6.hackit.mockspresso.reflect.TypeToken;
import org.easymock.EasyMock;
import org.easymock.Mock;
import org.powermock.api.easymock.PowerMock;
import org.powermock.api.easymock.annotation.MockNice;
import org.powermock.api.easymock.annotation.MockStrict;

import java.lang.annotation.Annotation;
import java.lang.reflect.Field;
import java.util.Arrays;
import java.util.List;

/**
 * A MockerConfig for PowerMock + EasyMock
 *
 * @deprecated This functionality is now exposed by the kotlin extension method `mockByPowerMock()` and its
 * JavaSupport counterpart {@link MockspressoEasyPowerMockPluginsJavaSupport#mockByPowerMock()}
 *
 * This class will be marked internal/protected in a future release
 */
@Deprecated
public class EasyPowerMockMockerConfig implements MockerConfig, MockerConfig.FieldPreparer, MockerConfig.MockMaker {

  @Override
  public MockMaker provideMockMaker() {
    return this;
  }

  @Override
  public FieldPreparer provideFieldPreparer() {
    return this;
  }

  @Override
  public List<Class<? extends Annotation>> provideMockAnnotations() {
    return Arrays.asList(
        Mock.class,
        org.powermock.api.easymock.annotation.Mock.class,
        MockStrict.class,
        MockNice.class);
  }

  @SuppressWarnings("unchecked")
  @Override
  public <T> T makeMock(TypeToken<T> typeToken) {
    // Using a strict mock by default here doesn't make sense for an auto-mocker
    T obj = PowerMock.<T>createNiceMock((Class<T>)typeToken.getRawType());

    // Since no one holds a reference to these mocks (besides the real object being
    // injected), they should be returned already in replay mode.
    EasyMock.replay(obj);
    return obj;
  }

  @Override
  public void prepareFields(Object objectWithMockFields) {
    try {
      prepareFieldsInternal(objectWithMockFields);
    } catch (IllegalAccessException e) {
      throw new RuntimeException(e);
    }
  }

  private void prepareFieldsInternal(Object objectWithMockFields) throws IllegalAccessException {
    List<Field> allFields = ReflectUtil.getAllDeclaredFields(objectWithMockFields.getClass());
    for (Field field : allFields) {
      if (!field.isAccessible()) {
        field.setAccessible(true);
      }

      if (field.get(objectWithMockFields) != null) {
        continue;
      }

      if (field.isAnnotationPresent(Mock.class) || field.isAnnotationPresent(org.powermock.api.easymock.annotation.Mock.class)) {
        field.set(objectWithMockFields, PowerMock.createMock(field.getType()));
      } else if (field.isAnnotationPresent(MockNice.class)) {
        field.set(objectWithMockFields, PowerMock.createNiceMock(field.getType()));
      } else if (field.isAnnotationPresent(MockStrict.class)) {
        field.set(objectWithMockFields, PowerMock.createStrictMock(field.getType()));
      }
    }
  }
}

File: mockspresso-easymock-powermock/src/main/java/com/episode6/hackit/mockspresso/easymock/powermock/EasyPowerMockMockerConfig.kt
package com.episode6.hackit.mockspresso.easymock.powermock

import com.episode6.hackit.mockspresso.api.MockerConfig
import com.episode6.hackit.mockspresso.api.MockerConfig.FieldPreparer
import com.episode6.hackit.mockspresso.reflect.ReflectUtil
import com.episode6.hackit.mockspresso.reflect.TypeToken
import org.easymock.EasyMock
import org.powermock.api.easymock.PowerMock
import java.lang.reflect.Field

/**
 * A MockerConfig for PowerMock + EasyMock
 */
internal class EasyPowerMockMockerConfig : MockerConfig, FieldPreparer, MockerConfig.MockMaker {
  override fun provideMockMaker(): MockerConfig.MockMaker = this
  override fun provideFieldPreparer(): FieldPreparer = this

  override fun provideMockAnnotations(): List<Class<out Annotation?>> = listOf(
      org.easymock.Mock::class.java,
      org.powermock.api.easymock.annotation.Mock::class.java,
      org.powermock.api.easymock.annotation.MockStrict::class.java,
      org.powermock.api.easymock.annotation.MockNice::class.java)


  @Suppress("UNCHECKED_CAST") // Using a strict mock by default here doesn't make sense for an auto-mocker
  override fun <T> makeMock(typeToken: TypeToken<T>): T = PowerMock.createNiceMock(typeToken.rawType as Class<T>).also {
    // Since no one holds a reference to these mocks (besides the real object being
    // injected), they should be returned already in replay mode.
    EasyMock.replay(it)
  }

  override fun prepareFields(objectWithMockFields: Any) {
    ReflectUtil.getAllDeclaredFields(objectWithMockFields.javaClass).forEach { field ->
      if (!field.isAccessible) field.isAccessible = true
      if (field[objectWithMockFields] != null) return@forEach
      when {
        field.isMock()       -> field[objectWithMockFields] = PowerMock.createMock(field.type)
        field.isNiceMock()   -> field[objectWithMockFields] = PowerMock.createNiceMock(field.type)
        field.isStrictMock() -> field[objectWithMockFields] = PowerMock.createStrictMock(field.type)
      }
    }
  }
}

private fun Field.isMock() = isAnnotationPresent(org.easymock.Mock::class.java) || isAnnotationPresent(org.powermock.api.easymock.annotation.Mock::class.java)
private fun Field.isNiceMock() = isAnnotationPresent(org.powermock.api.easymock.annotation.MockNice::class.java)
private fun Field.isStrictMock() = isAnnotationPresent(org.powermock.api.easymock.annotation.MockStrict::class.java)
